src/main/java/org/candlepin/pki/impl/BouncyCastlePKIUtility.java

```diff
@@ -0,0 +1,528 @@
+     *  class
+     *
+     * @return
+     *  an instance of the given key class representing the key data provided
+     */
+    private <T extends Key> T deserializeKey(byte[] bytes, Class<T> keyClass)
+        throws ClassNotFoundException, IOException {
+
+        if (bytes == null) {
+            throw new IOException("no data to deserialize");
+        }
+
+        try (ObjectInputStream ostream = new ObjectInputStream(new ByteArrayInputStream(bytes))) {
+            Object obj = ostream.readObject();
+
+            if (!keyClass.isInstance(obj)) {
+                throw new IOException("incorrect object parsed from key data: " + obj.getClass());
+            }
+
+            return (T) obj;
+        }
+    }
+
+    /**
+     * Generates a PublicKey instance from the provided key data and algorithm.
+     *
+     * @param keydata
+     *  the X509 encoded public key data from which to generate a PublicKey instance
+     *
+     * @param algorithm
+     *  the algorithm used to generate the key
+     *
+     * @return
+     *  a PublicKey instance
+     */
+    private PublicKey generatePublicKey(byte[] keydata, String algorithm)
+        throws NoSuchAlgorithmException, InvalidKeySpecException {
+
+        KeyFactory factory = KeyFactory.getInstance(algorithm, BouncyCastleProviderLoader.getProvider());
+        X509EncodedKeySpec spec = new X509EncodedKeySpec(keydata, algorithm);
+
+        return factory.generatePublic(spec);
+    }
+
+    /**
+     * Generates a PrivateKey instance from the provided key data and algorithm.
+     *
+     * @param keydata
+     *  the PKCS8 private key data from which to generate a PrivateKey instance
+     *
+     * @param algorithm
+     *  the algorithm used to generate the key
+     *
+     * @return
+     *  a PrivateKey instance
+     */
+    private PrivateKey generatePrivateKey(byte[] keydata, String algorithm)
+        throws NoSuchAlgorithmException, InvalidKeySpecException {
+
+        KeyFactory factory = KeyFactory.getInstance(algorithm, BouncyCastleProviderLoader.getProvider());
+        PKCS8EncodedKeySpec spec = new PKCS8EncodedKeySpec(keydata, algorithm);
+
+        return factory.generatePrivate(spec);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public KeyPair generateKeyPair() throws KeyException {
+        try {
+            KeyPairGenerator keyGen = KeyPairGenerator.getInstance(KEY_ALGORITHM);
+            keyGen.initialize(KEY_SIZE);
+            KeyPair keypair = keyGen.generateKeyPair();
+            return keypair;
+        }
+        catch (NoSuchAlgorithmException e) {
+            throw new KeyException(e);
+        }
+    }
+
+    /**
+     * Fetches the encoded form of the specified private key
+     *
+     * @param privateKey
+     *  the private key from which to fetch the encoded form
+     *
+     * @return
+     *  the encoded form of the given private key
+     */
+    private byte[] getKeyEncoding(PrivateKey privateKey) throws KeyException {
```

Comment:
This method is still using the old wrap->unwrap flow to get the key data as bytes that JSS used to impose on us. Everything below 495 is likely unused entirely under BouncyCastle.